0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+   
+
+Here is why the [Detailed Design](#detailed-design) section suggests "transposing" the signature and using array-based arguments with broadcasting:
+
+ 1. Transposing the signature will make it obvious what the primitive intends to do with each circuit.
+ 2. Transposing the signature will let us introduce and reason about the notion  "primitive unit of work", for which this RFC proposes the name "task".
+ 3. Array-based arguments will let users assign operational meaning to each axis (this axis is for twirling, that axis is for basis changes, etc.).
+ 4. Broadcasting rules will let users choose how to combine parameter value sets with observables in different ways, such as
+    1. use one observable for all of N parameter value sets
+    2. use one parameter value set for all of N observables
+    3. zip N parameter value sets against N observables
+    4. take all NxM combinations of N parameter value sets with M observables
+    5. etc.
+
+## User Benefit <a name="user-benefit"></a>
+
+All users of the primitives stand to benefit from this proposal.
+Immediately, it will enable sophisticated and convenient workflows for power users though arrays and broadcasting.
+However, standard broadcasting rules are such that the 0D and 1D cases will feel natural to existing users---0D is essentially what we already have, and 1D will be perceived as a simple-yet-welcome bonus.
+
+Moreover, the interface changes in this proposal will enable specific primitive implementations to enhance performance through reduced bandwidth on the runtime, compatibility with fast parameter binding, and multiplexing qubit-wise commuting observables.
+
+## Design Proposal <a name="design-proposal"></a>
+
+We will lead the design proposal with an example:
+
+```python
+# construct two circuits, the first with 2039 Parameters, the second with no Parameters
+circuit1 = QuantumCircuit(9)
+...
+circuit2 = QuantumCircuit(2)
+...
+
+# specify 128 different parameter value sets for circuit1, in a 32x4 shape
+parameter_values1 = np.random.uniform(size=(32, 4, 2039))
+# specify 4 observables to measure for circuit 1
+observables1 = ["ZZIIIIIII", "IZXIIIIII", "IIIIIIYZI", "IZXIIYIII"]
+
+# specify 18 observables to measure for circuit2
+observables2 = [Pauli("XYZIIII"), ..., SparsePauliOp({"YYIIIII": 0.2})]
+
+# invoke an estimator
+estimator = Estimator()
+job = estimator.run([
+  (circuit1, observables1, parameter_values1),
+  (circuit2, observables2)
+])
+
+# wait for the result
+job.result()
+
+# get a bundle of results for every task
+>> EstimatorResult(
+>>     TaskResult<{evs: ndarray<32, 4>, stds: ndarray<32, 4>}, local_metadata>, 
+>>     TaskResult<{evs: ndarray<18>, stds: ndarray<18>}, local_metadata>, 
```

Comment:
Are TaskResults here in some for of list/array corresponding to the broadcast shape?

Reply:
There is one TaskResult for each Task. Each TaskResult is shaped like the corresponding task.

Reply:
I think my question was more that the TaskResults seemed to be individual parameters to the EstimatorResult object, followed by some global metadata. Below estimator.run (line 164) shows run returning a list of ResultTasks. The input to the run() just above is a list of tuples- the two "tasks". Maybe this is more illustrative here than actual in that it shows two things as well, and maybe they end up in a list or some Iterable part of result.